### ISSUE TITLE
Encoder decoder

### ISSUE THREAD

**Issue by him4318** [Hannibal046/nanoRWKV#5]
Hi 
I want to train an OCR model using RWKV but I am stuck at data formatting.
I tried to prepend the image tokens with text tokens and send it through RWKV.
After that I calculated the loss with token shifting:
 ```
#while creating the targets I made all image tokens as zero so that the loss is only calculated on text tokens
targets = targets[:,1:] 
logits,_ = rwkv(token_embs)
logits = logits[:,:-1,:]
#goes on to calulate the loss with Cross entropy ignoring the index=0

```
But the model is not learning. Is there a way where we can make a seq2seq using RWKV?
where encoder could be image features from SWIN and we can append or add these features to RWKV while training.

**Comment by Hannibal046** [comment]
Hi, what do you mean by not learning? There is no available seq2seq model with RWKV as decoder. But i believe it would easy to hack the current nanoRWKV to it.

**Comment by him4318** [comment]
It only learns to predict EOS tokens. 
I am confused how to hack it so that I add cross attention to the decoder.

**Comment by Hannibal046** [comment]
I don't know how you exactly model RWKV for OCR. But RWKV is essentially an LM without attention. Training RWKV is equivalent to training a Transformer-based LM. Did you successfully run Transformer for OCR?

AFAIK, there is no attetnion-RWKV mixing model, if you want to simply add another cross-attention module in RWKV, you could modify from here:
https://github.com/Hannibal046/nanoRWKV/blob/3f745a4846e71f6462db2c12e498808dfe155d1b/modeling_rwkv.py#L222-L227

**Comment by him4318** [comment]
I did train the transformer for ocr but it was a full transformer where the memory(image features) is used to do the cross attention https://github.com/him4318/Transformer-ocr.

For RWKV what I am doing is prepend the image feature vector to the token embeddings.
`token_embs = torch.cat([image_embs, token_embs], dim=1) `
then passed the token_embs to RWKV LM.

I am not much aware about the arch of RWKV so don't have the idea what parts need to be changed so that each decoder token can attend to the feature vector while predicting. 
`Test it on bidirectional & MLM tasks, and image & audio & video tokens. I think RWKV can support Encoder-Decoder via this: for each decoder token, use a learned mixture of [decoder previous hidden state] & [encoder final hidden state]. Hence all decoder tokens will have access to the encoder output.`

This was mentioned by the author of the original RWKV but couldn't get myself to build it.

**Comment by Hannibal046** [comment]
I understand. But this is out of the scope of this repo. Maybe we could wait authors of RWKV to implement this.
